library/common/http/client.cc

```diff
@@ -479,6 +480,16 @@ void Client::sendHeaders(envoy_stream_t stream, envoy_headers headers, bool end_
   if (direct_stream) {
     ScopeTrackerScopeState scope(direct_stream.get(), scopeTracker());
     RequestHeaderMapPtr internal_headers = Utility::toRequestHeaders(headers);
+
+    // This is largely a check for the android platform: is_cleartext_permitted
+    // is a no-op for other platforms.
+    if (internal_headers->getSchemeValue() != "https" &&
+        !is_cleartext_permitted(internal_headers->getHostValue())) {
+      direct_stream->request_decoder_->sendLocalReply(
+          Http::Code::BadRequest, "Cleartext is not permitted", nullptr, absl::nullopt, "");
```

Comment:
Rather than issuing the direct reply, could we simply call the on_error callback here?

Reply:
I guess the current setup will result in the filter chain being created, whereas firing the callback directly wouldn't(?). I don't know if that's desirable or not. Maybe it is, if someone is listening for errors in a filter?

But I suppose I'm fine either way, as long as we're consistent about we handle this sort of thing for any other cases moving forward.